src/main/java/me/contaria/seedqueue/mixin/client/WorldGenerationProgressTrackerMixin.java

```diff
@@ -0,0 +1,105 @@
+import org.jetbrains.annotations.Nullable;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.Optional;
+
+@Mixin(WorldGenerationProgressTracker.class)
+public abstract class WorldGenerationProgressTrackerMixin implements SQWorldGenerationProgressTracker {
+    @Shadow @Final
+    private Long2ObjectOpenHashMap<ChunkStatus> chunkStatuses;
+    @Shadow
+    private ChunkPos spawnPos;
+    @Shadow @Final
+    private int radius;
+
+    @Shadow @Final private WorldGenerationProgressLogger progressLogger;
+    @Unique
+    private long freezeTime = -1;
+    @Unique
+    @Nullable
+    WorldGenerationProgressTracker frozenCopy = null;
+
+    @Unique
+    private int progressOverride = -1; // Overrides the returned progress percentage if equal to anything other than -1
+
+    @Inject(
+            method = "<init>",
+            at = @At("TAIL")
+    )
+    private void markFreezeTimeStart(int radius, CallbackInfo ci) {
+        long chunkMapFreezingTime = SeedQueue.config.chunkMapFreezing;
+        if (chunkMapFreezingTime != -1 && this.freezeTime == -1) {
+            this.makeFrozenCopyAfter(chunkMapFreezingTime);
+        }
+    }
+
+    @Inject(
+            method = "setChunkStatus",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/server/WorldGenerationProgressLogger;setChunkStatus(Lnet/minecraft/util/math/ChunkPos;Lnet/minecraft/world/chunk/ChunkStatus;)V",
+                    shift = At.Shift.BEFORE
+            )
+    )
+    private void onSetChunkStatus(ChunkPos pos, ChunkStatus status, CallbackInfo ci) {
+        if (this.frozenCopy == null && this.isPastFreezingTime()) {
+            this.makeFrozenCopy();
+        }
+    }
+
+    @Inject(
+            method = "getProgressPercentage",
+            at = @At("HEAD"),
+            cancellable = true
+    )
+    private void replaceProgressPercentage(CallbackInfoReturnable<Integer> cir) {
+        if (this.progressOverride != -1) {
+            cir.setReturnValue(this.progressOverride);
+        }
+    }
+
+    @Unique
+    private void makeFrozenCopy() {
+        this.frozenCopy = new WorldGenerationProgressTracker(this.radius - ChunkStatus.getMaxTargetGenerationRadius()); // This will trigger a makeFrozenCopyAfter inside the frozen copy itself which could lead to further recursion, but as long as setChunkStatus isn't called, this will never be an issue.
+        ((WorldGenerationProgressTrackerMixin) (Object) this.frozenCopy).setAsFrozenCopy(this.chunkStatuses, this.spawnPos, this.progressLogger.getProgressPercentage());
```

Comment:
Please change this to something like `this.frozenCopy = new WorldGenerationProgressTracker(...).setAsFrozenCopy(...)` since this is accessed from multiple threads and it might technically be able to get the frozen copy before it got its copy

Reply:
![image](https://github.com/user-attachments/assets/59004bd9-79a2-4734-8e81-108f93bec28f)
lol

Reply:
Update MCDev and IDEA or just return `WorldGenerationProgressTracker` in `setAsFrozenCopy`